server/src/main/java/org/elasticsearch/search/fetch/subphase/MatchedQueriesFetchSubPhase.java

```diff
@@ -18,82 +18,69 @@
-        }
 
-        Map<String, Query> namedQueries = new HashMap<>(context.parsedQuery().namedFilters());
-        if (context.parsedPostFilter() != null) {
-            namedQueries.putAll(context.parsedPostFilter().namedFilters());
-        }
+        Query q = getQuery(context);
+        Weight w = context.searcher().createWeight(context.searcher().rewrite(q), ScoreMode.COMPLETE_NO_SCORES, 1);
+        SearchHit[] sortedHits = hits.clone();
+        Arrays.sort(sortedHits, Comparator.comparingInt(SearchHit::docId));
+        for (SearchHit hit : sortedHits) {
+            int ord = ReaderUtil.subIndex(hit.docId(), context.searcher().getTopReaderContext().leaves());
+            LeafReaderContext ctx = context.searcher().getTopReaderContext().leaves().get(ord);
+            int doc = hit.docId() - ctx.docBase;
+            Matches m = w.matches(ctx, doc);
+            if (m == null) {
+                continue;
+            }
 
-        try {
-            for (Map.Entry<String, Query> entry : namedQueries.entrySet()) {
-                String name = entry.getKey();
-                Query query = entry.getValue();
-                int readerIndex = -1;
-                int docBase = -1;
-                Weight weight = context.searcher().createWeight(context.searcher().rewrite(query), ScoreMode.COMPLETE_NO_SCORES, 1f);
-                Bits matchingDocs = null;
-                final IndexReader indexReader = context.searcher().getIndexReader();
-                for (int i = 0; i < hits.length; ++i) {
-                    SearchHit hit = hits[i];
-                    int hitReaderIndex = ReaderUtil.subIndex(hit.docId(), indexReader.leaves());
-                    if (readerIndex != hitReaderIndex) {
-                        readerIndex = hitReaderIndex;
-                        LeafReaderContext ctx = indexReader.leaves().get(readerIndex);
-                        docBase = ctx.docBase;
-                        // scorers can be costly to create, so reuse them across docs of the same segment
-                        ScorerSupplier scorerSupplier = weight.scorerSupplier(ctx);
-                        matchingDocs = Lucene.asSequentialAccessBits(ctx.reader().maxDoc(), scorerSupplier);
-                    }
-                    if (matchingDocs.get(hit.docId() - docBase)) {
-                        matchedQueries[i].add(name);
-                    }
-                }
+            for (NamedQuery.NamedMatches nm : NamedQuery.findNamedMatches(m)) {
+                hit.addMatchedQuery(nm.getName());
             }
-            for (int i = 0; i < hits.length; ++i) {
-                hits[i].matchedQueries(matchedQueries[i].toArray(new String[matchedQueries[i].size()]));
+        }
+    }
+
+    private Query getQuery(SearchContext context) {
+        if (context.isNested()) {
+            BooleanQuery.Builder builder = new BooleanQuery.Builder();
+            for (Query q : context.parsedQuery().getNestedQueries()) {
```

Comment:
Shouldn't we check that the parent nested query matches the nested path of the current context ?